extensions/interactions/GraphInput/directives/graph-viz.component.ts

```diff
@@ -0,0 +1,649 @@
+import { GraphAnswer } from 'interactions/answer-defs';
+import { InteractionsExtensionsConstants } from 'interactions/interactions-extension.constants';
+import { PlayerPositionService } from 'pages/exploration-player-page/services/player-position.service';
+import { Subscription } from 'rxjs';
+import { DeviceInfoService } from 'services/contextual/device-info.service';
+import { FocusManagerService } from 'services/stateful/focus-manager.service';
+import { UtilsService } from 'services/utils.service';
+import { EdgeCentre, GraphDetailService } from './graph-detail.service';
+import { downgradeComponent } from '@angular/upgrade/static';
+
+const debounce = (delay: number = 5): MethodDecorator => {
+  return function(
+      target: unknown, propertyKey: string, descriptor: PropertyDescriptor) {
+    const original = descriptor.value;
+    const key = `__timeout__${propertyKey}`;
+
+    descriptor.value = function(...args) {
+      clearTimeout(this[key]);
+      this[key] = setTimeout(() => original.apply(this, args), delay);
+    };
+    return descriptor;
+  };
+};
+
+interface GraphButton {
+  text: string;
+  description: string;
+  mode: number
+}
+
+interface GraphOption {
+  text: string;
+  option: string;
+}
+@Component({
+  selector: 'graph-viz',
+  templateUrl: './graph-viz.component.html',
+  styleUrls: []
+})
+export class GraphVizComponent implements OnInit, AfterViewInit {
+  @Input() graph: GraphAnswer;
+  @Input() canAddVertex: boolean;
+  @Input() canDeleteVertex: boolean;
+  @Input() canMoveVertex: boolean;
+  @Input() canEditVertexLabel: boolean;
+  @Input() canAddEdge: boolean;
+  @Input() canDeleteEdge: boolean;
+  @Input() canEditEdgeWeight: boolean;
+  @Input() interactionIsActive: boolean;
+  @Input() canEditOptions: boolean;
+  isMobile: boolean = false;
+  helpText: string = '';
+  _MODES = {
+    MOVE: 0,
+    ADD_EDGE: 1,
+    ADD_VERTEX: 2,
+    DELETE: 3
+  };
+  // Styling functions.
+  DELETE_COLOR = 'red';
+  HOVER_COLOR = 'aqua';
+  SELECT_COLOR = 'orange';
+  DEFAULT_COLOR = 'black';
+  state = {
+    currentMode: this._MODES.MOVE,
+    // Vertex, edge, mode button, label currently being hovered over.
+    hoveredVertex: null,
+    hoveredEdge: null,
+    hoveredModeButton: null,
+    // If in ADD_EDGE mode, source vertex of the new edge, if it
+    // exists.
+    addEdgeVertex: null,
+    // Currently dragged vertex.
+    currentlyDraggedVertex: null,
+    // Selected vertex for editing label.
+    selectedVertex: null,
+    // Selected edge for editing weight.
+    selectedEdge: null,
+    // Mouse position in SVG coordinates.
+    mouseX: 0,
+    mouseY: 0,
+    // Original position of dragged vertex.
+    vertexDragStartX: 0,
+    vertexDragStartY: 0,
+    // Original position of mouse when dragging started.
+    mouseDragStartX: 0,
+    mouseDragStartY: 0
+  };
+  selectedEdgeWeightValue: number | string;
+  buttons: GraphButton[] = [];
+  private vizContainer: SVGSVGElement[];
```

Comment:
Just curious, why is the 'SVG' part twice in the type name?

Reply:
Maybe this might help: https://developer.mozilla.org/en-US/docs/Web/API/SVGSVGElement